### ISSUE TITLE
Skip to content link for accessibility

### ISSUE THREAD

**Issue by ghost** [h5bp/html5-boilerplate#1692]
Hey,

Long time user, first time issuer :)

As a source of best practices, should H5BP contain a "skip to content" link. It'll mean we'll have to add a tag that can be skipped to, but I think its a good idea, because accessibility is important and often missing. Might make people think more about how continue to build their sites as well.

Thoughts?

Bit of info here on it: https://www.webaccessibility.com/best_practices.php?technology_platform_id=13
Done here: http://www.w3.org/WAI/

**Comment by patrickkettner** [comment]
I am very interested in the concept. One thing to keep in mind, however, is that it won't work cross browser without javascript. chrome (and safari last time I checked) does not update tabfocus on `hashChange`.

**Comment by WebNinjrDesign** [comment]
Just seen this as well: http://a11ywins.tumblr.com/post/116335338923/nytimes-coms-skip-link-design

**Comment by Nettsentrisk** [comment]
This is going to vary from site to site, and if you do your accessibility right, this isn't actually needed. Screen reader users don't typically use nor need this. 

WebAIM wrote about the alternatives to using this "hack" as they call it: http://webaim.org/techniques/skipnav/#alternatives

One of their surveys showed quite a mixed bag relating to the use of this feature: http://webaim.org/projects/screenreadersurvey4/#skip

Most screen reader users use headings, landmarks or other ways to find what they're looking for on a page: http://webaim.org/projects/screenreadersurvey4/#finding

The only group left is keyboard-only users who don't use a screen reader. How many of them are there and is this feature useful to them?

I've wrestled with this topic and I can't really land on a solid answer, indicating to me that it's not ripe for a "boilerplate".

**Comment by Tigt** [comment]
In IE, in order to give tab focus to the jumped-to element, [the element needs to be "accessible."](http://jimthatcher.com/news-070607.htm) This worked with named anchors (`<a name="jump-to-me"></a>`), but has since become obsolete.

One can hack an element to "become accessible" without disrupting the tab flow by adding `tabindex="-1"` to it.

[This bug is *still around,* even in Edge.](https://connect.microsoft.com/IE/Feedback/Details/1111056) It's kind of pitiful, really. Please do vote on it.

I agree this sort of thing *should* be standard, and the closest thing possible to that is landmark navigation. But for the 63.5% of users in that WebAIM survey who use skip links at least "sometimes," one needs something like this:

```html
<a href="#main-content">Skip to main content</a>

<!-- your site navigation and other folderol -->

<main id="main-content" tabindex="-1">
    <!-- the thing the user actually came to the page for -->
</main>
```

Hacky, and a bit weird, but necessary.

**Comment by Nettsentrisk** [comment]
They use it because/if it's there, but there's a lot of evidence that they would not be hampered in any real way by it not being there. Due to the fact that every site is going to solve this differently (although a boilerplate would certainly make a dent), it's also something that isn't reliable for screen reader users. In cases like that, it's best to code your site so that their own tools, which they are most familiar with, work well. Sticking to the WCAG rules about structure, headings, etc. will let their own tools present them a variety of ways to jump around the page to what they want.

**Comment by scottaohara** [comment]
I would agree that this is a good feature to add to the boilerplate, as it may not be as useful for people who use screen readers, but it can be helpful for anyone that uses a keyboard as their primary method of navigation. Accessibility isn't just about visual users vs blind users, after all.

If more evidence is needed that it's a good idea, [here's the WCAG rule that says a skip link should be included to meet Level A compliance](https://www.w3.org/WAI/WCAG20/quickref/#navigation-mechanisms-skip)

**Comment by Nettsentrisk** [comment]
> If more evidence is needed that it's a good idea, here's the WCAG rule that says a skip link should be included to meet Level A compliance

It's important to distinguish between "should be included" and "can be included". WCAG gives a list of different ways this requirement can be met, a skip link is just one way of doing it. A boilerplate should probably not be the place to be opinionated when different solutions are available.

**Comment by scottaohara** [comment]
Actually, the success criteria is pretty much reliant on a skip link being present, and its benefit is to provide a consistent, __quick__ way for users to access the main content of a site. 

Using external tools (like the screen reader functionality you reference) does not meet success criteria because that would mean people that use keyboard controls to navigate a document also _have_ to use a screen reader.   And while they may be a 'hack', even the WebAim articles you reference conceded that there is no native way to provide this functionality, so this hack is actually quite useful.

Additionally there is even benefit to screen reader users as it's also quicker to simply tab and activate the first link on a page (the skip link) than it is to go to a page, open landmark or heading navigation controls, parse through the list and then make a selection.  

> A boilerplate should probably not be the place to be opinionated...

This argument is pretty weak in that it there are opinions already in place in what should be included in the H5BP (jQuery for one), and I would submit that there is more benefits to including this feature in the boilerplate than your opinion to keep it out.

**Comment by Nettsentrisk** [comment]
> Actually, the success criteria is pretty much reliant on a skip link being present

The skip link technique (G1) is only 1 of 10 listed sufficient techniques, ARIA11 and H69 being the main alternatives for general use. Any of these techniques fulfill the requirement, keyboard-only users be damned.

> its benefit is to provide a consistent, quick way for users to access the main content of a site.

I would agree - if the majority of web sites on the Internet used H5BP, or used this pattern, neither of which is the case. The end result is a pattern which many users don't recognize, which will look and function differently on every single implementation even if it's provided in a boilerplate. A popular parallel is the "hamburger" menu. "Consistent" is not the result.

I agree that the keyboard-only users are at a loss here, but at some point it's up to browser vendors to provide this functionality in a way that works the same on every single web page, and not leaving this up to each individual site to create their own version of a pattern that's flawed from the start.

We have the <main> element, browser vendors should let keyboard-only users and screen readers jump to it automatically. Instead of implementing hacks in web pages to circumvent this problem, we should be pressuring browser vendors to implement it. And yes, that's my opinion. :)

**Comment by scottaohara** [comment]
> keyboard-only users be damned

well alright then, bye bye :)

**Comment by Nettsentrisk** [comment]
You can complain to the WCAG authors, I'm just the messenger :)

**Comment by roblarsen** [comment]
I'm going revisit this for the doc sweep. Closing.
